File: go/loss/mse.go
package loss

import (
	shapes "github.com/flygerian/shapes"
)

// Mse computes the mean squared error loss between ground truth and predictions.
// yGround and yPred should be WrappedTensors with the same shape.
// Returns: sum((yPred - yGround)^2) reduced to a scalar.
func Mse(c shapes.Context, yGround shapes.Tensor, yPred shapes.Tensor) func(shapes.Tensor) shapes.Tensor {
	return func(t shapes.Tensor) shapes.Tensor {
		_ = t

		fusedCtx := c.Forward(shapes.WithInputs(yGround, yPred))

		// Compute: (yPred - yGround)^2
		diff := yPred.Minus(fusedCtx, yGround)
		se := diff.Pow(fusedCtx, 2)

		// Sum over all dimensions to reduce to scalar
		result := se
		shape := result.Shape()
		for i := len(shape) - 1; i >= 0; i-- {
			if shape[i] > 1 {
				result = result.Sum(fusedCtx, uint32(i))
			}
		}

		fusedCtx.Finish(
			shapes.WithResult(result),
			shapes.WithOpType(shapes.OpMse),
			shapes.WithBackward(mseBackward),
		)

		return result
	}
}

// mseBackward computes gradients for MSE loss.
// loss = sum((yPred - yGround)^2)
// ∂L/∂yPred = 2*(yPred - yGround)
// ∂L/∂yGround = -2*(yPred - yGround)
func mseBackward(c shapes.Context, node shapes.ComputationGraphNode) {
	ctx := c.Backward()
	defer ctx.Finish()

	yGround := node.Inputs()[0]
	yPred := node.Inputs()[1]

	// ∂L/∂yPred = upstream_grad * 2*(yPred - yGround)
	diff := yPred.Minus(ctx, yGround)
	two := shapes.Float(ctx, diff.Shape(), 2.0)
	localGrad := two.Times(ctx, diff)
	gradYPred := node.Grad().Times(ctx, localGrad)
	reducedYPred := shapes.ReduceBroadcast(ctx, yPred, gradYPred.(shapes.GradTensor))
	yPred.Grad().Accumulate(ctx, reducedYPred)

	// ∂L/∂yGround = upstream_grad * -2*(yPred - yGround)
	negLocalGrad := localGrad.Negate(ctx)
	gradYGround := node.Grad().Times(ctx, negLocalGrad)
	reducedYGround := shapes.ReduceBroadcast(ctx, yGround, gradYGround.(shapes.GradTensor))
	yGround.Grad().Accumulate(ctx, reducedYGround)
}

File: base/nn/mse.c
#include "result/result.h"
#include "shapes.h"
#include <stddef.h>
#include "tensor_internal.h"
#include "utils_lib/array.h"

void mseBackward(Context *ctx, Tensor *tensor) {
  PANIC_IF(ctx == NULL || tensor == NULL || tensor->inputs == NULL || tensor->grad == NULL, ERR_NULL_TENSOR_PROVIDED);
  PANIC_IF(tensor->inputs->size < 2, ERR_NULL_TENSOR_PROVIDED);

  Tensor *yGround = shapes_Array_TensorIdx(tensor->inputs, 0);
  Tensor *yPred = shapes_Array_TensorIdx(tensor->inputs, 1);

  PANIC_IF(yGround == NULL, ERR_NULL_TENSOR_PROVIDED);
  PANIC_IF(yPred == NULL, ERR_NULL_TENSOR_PROVIDED);

  // ∂L/∂yPred = upstream_grad * 2*(yPred - yGround)
  Tensor diff = shapes_Subtract(ctx, yPred, yGround);
  Tensor two = shapes_Make_FloatTensor(ctx, diff.shape, 2.0);
  Tensor localGrad = shapes_Multiply(ctx, &two, &diff);
  Tensor gradYPred = shapes_Multiply(ctx, tensor->grad, &localGrad);
  Tensor reducedGradYPred = shapes_ReduceBroadcast(ctx, yPred, &gradYPred);
  shapes_AddInPlace(ctx, yPred->grad, &reducedGradYPred);

  // ∂L/∂yGround = upstream_grad * -2*(yPred - yGround)
  Tensor negLocalGrad = shapes_Negate(ctx, &localGrad);
  Tensor gradYGround = shapes_Multiply(ctx, tensor->grad, &negLocalGrad);
  Tensor reducedGradYGround = shapes_ReduceBroadcast(ctx, yGround, &gradYGround);
  shapes_AddInPlace(ctx, yGround->grad, &reducedGradYGround);
}

Tensor shapesnn_Mse(Context *ctx, Tensor *yGround, Tensor *yPred) {
  Tensor diffVal = shapes_Subtract(ctx, yPred, yGround);

  Tensor *loss = allocate(ctx->memory, sizeof(Tensor));
  PANIC_IF(loss == NULL, ALLOCATION_FAILED);
  *loss = shapes_Pow(ctx, &diffVal, 2);

  for (dim_t i = 0; i < loss->shape.numOfDims; i++) {
    dim_t dim = loss->shape.dims[i];
    if (dim > 1) {
      Tensor *newLoss = allocate(ctx->memory, sizeof(Tensor));
      PANIC_IF(newLoss == NULL, ALLOCATION_FAILED);
      *newLoss = shapes_Sum(ctx, loss, i);
      loss = newLoss;
    }
  }

  loss->inputs = MakeArray(ctx->memory, sizeof(Tensor), 2);
  shapes_Array_AppendTensor(loss->inputs, yGround);
  shapes_Array_AppendTensor(loss->inputs, yPred);

  loss->opType = OP_MSE;
  return *loss;
}
